Add transpose_kernel argument to ConvTranspose.

mathisgerdes opened the pull request:
This is a proposal for #2577, adding a `transpose_kernel` option to `ConvTranspose` such that the previous behavior is maintained by default.

- Add an additional `transpose_kernel` parameter to `ConvTranspose` which is passed to `jax.lax.conv_transpose`
- For  `padding='CIRCULAR'`: To maintain the previous convention of `ConvTranspose`, the alignment of the kernel is changed depending on `transpose_kernel`.
   ```python
   # Divide the padding equally between left and right. The choice to put
   # "+1" on the left (or on the right) represents a convention for
   # aligning even-sized kernels.
   if self.transpose_kernel:
     # match convention in Conv to achieve the proper transpose
     total_pad = [
        (size_diff // 2, (size_diff + 1) // 2) for size_diff in size_diffs
     ]
   else:
     # keep the previous convention for backward compatibility
     total_pad = [
        ((size_diff + 1) // 2, size_diff // 2) for size_diff in size_diffs
     ]
   ```

cgarciae commented on flax/linen/linear.py:
```diff
@@ -688,12 +695,17 @@ def __call__(self, inputs: Array) -> Array:
           for y_dim, x_dim in zip(y.shape[1:-1], scaled_x_dims)
       ]
-      # Divide the padding equaly between left and right. The choice to put
+      # Divide the padding equally between left and right. The choice to put
```
Maybe we could move this comment to the `else` branch and have a similar one for the positive branch?

mathisgerdes replied:
I agree, the comments there could be clearer. How about this modified version, with clearer comments for both cases?

cgarciae reviewed (approved):
Thanks @mathisgerdes for doing this! Exposing `transpose_kernel` is a great idea.

marcvanzee commented:
@mathisgerdes @cgarciae this PR has build failures so I have removed the "pull ready" label. Could you please fix those first?

mathisgerdes commented:
> @mathisgerdes @cgarciae this PR has build failures so I have removed the "pull ready" label. Could you please fix those first?

I believe those were problems with the previous commit, not any changes here. I've rebased to the latest commit.

mathisgerdes commented:
Is there still something to be done here, @cgarciae, @andsteing?

cgarciae reviewed (approved):
Sorry for the delay. LGTM. 
Thanks for doing this!
